cmd/juju/cloud/detectcredentials.go

```diff
@@ -122,18 +154,54 @@ type discoveredCredential struct {
 	isNew            bool
 }
 
+func (c *detectCredentialsCommand) credentialsAPI() (CredentialAPI, error) {
+	var err error
+	root, err := c.NewAPIRoot(c.Store, c.controllerName, "")
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return apicloud.NewClient(root), nil
+}
+
 func (c *detectCredentialsCommand) allClouds() (map[string]jujucloud.Cloud, error) {
 	clouds, _, err := jujucloud.PublicCloudMetadata(jujucloud.JujuPublicCloudsPath())
 	if err != nil {
 		return nil, err
 	}
+	builtinClouds, err := common.BuiltInClouds()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for k, v := range builtinClouds {
+		clouds[k] = v
+	}
 	personalClouds, err := jujucloud.PersonalCloudMetadata()
 	if err != nil {
 		return nil, err
 	}
 	for k, v := range personalClouds {
 		clouds[k] = v
 	}
+	if !c.Local {
+		// If there is a cloud definition for the same cloud both
+		// on the controller and on the client and they conflict,
+		// we want definition from the controller to take precedence.
+		client, err := c.credentialAPIFunc()
+		if err != nil {
+			return nil, err
+		}
+		defer client.Close()
+
+		remoteUserClouds, err := client.Clouds()
+		if err != nil {
+			return nil, err
+		}
+		c.remoteClouds = map[string]jujucloud.Cloud{}
+		for k, v := range remoteUserClouds {
+			clouds[k.Id()] = v
```

Comment:
do we want to warn if the personal cloud(same name but different content) gets overwritten by remote cloud?

Reply:
Not in this case. Here we do want to be consistent with all other places where we are dealing with both local and remote clouds. To say anything additional here would be too noisy.

However, in our f-2-f discussion, this does bring an in interesting point about what happens when local and remote clouds mismatch? what happens when remote clouds get updated? What happens if it gets updated with  outdated information? All interesting questions that need to be answered separately (outside of this PR) and consistently (across all functionalities that deal with local/remote clouds as well as deal with cloud management).